### PR TITLE
[#3264] feat(spark-connector): Support Iceberg time travel in SQL queries

### PR DESCRIPTION
**Pull request by caican00**
### What changes were proposed in this pull request?

Support Iceberg time travel in SQL queries

### Why are the changes needed?
supports time travel in SQL queries using `TIMESTAMP AS OF`, `FOR SYSTEM_TIME AS OF` or `VERSION AS OF`, `FOR SYSTEM_VERSION AS OF` clauses.

Fix: https://github.com/datastrato/gravitino/issues/3264

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
New ITs.

### CODE REVIEW

**Comment by caican00** [comment]
Finally, I still choose to implement Iceberg `time travel` by overriding `newScanBuilder`, for the following reasons:

1. Although `SparkIcebergTable` extended `SparkTable`, it still needs to initialize its member variables, such as `snapshotId` or `branch`, before it can directly reuse the `newScanBuilder` implementation of `SparkTable`.

2. However, initializing `snapshotId` or `branch` is difficult, not as easy as initializing `refreshEagerly`, because it is difficult to determine `snapshotId` or `branch` is initialized in the real sparkTable. Therefore, it is difficult to selectively initialize `snapshotId` or `branch` through the `super` method.

In this case, users specify the version for time travel, but the version may be snapshotId or branch name.
https://github.com/apache/iceberg/blob/2058053b0c6e5b1c7e91fa029162f22d109aafb1/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java#L195-L199

3. Therefore, overriding `newScanBuilder` is more convenient and does not introduce too much maintenance burden.

**Comment by caican00** [comment]
> Finally, I still choose to implement Iceberg `time travel` by overriding `newScanBuilder`, for the following reasons:
> 
> 1. Although `SparkIcebergTable` extended `SparkTable`, it still needs to initialize its member variables, such as `snapshotId` or `branch`, before it can directly reuse the `newScanBuilder` implementation of `SparkTable`.
> 2. However, initializing `snapshotId` or `branch` is difficult, not as easy as initializing `refreshEagerly`, because it is difficult to determine `snapshotId` or `branch` is initialized in the real sparkTable. Therefore, it is difficult to selectively initialize `snapshotId` or `branch` through the `super` method.
> 
> In this case, users specify the version for time travel, but the version may be snapshotId or branch name. https://github.com/apache/iceberg/blob/2058053b0c6e5b1c7e91fa029162f22d109aafb1/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java#L195-L199
> 
> 3. Therefore, overriding `newScanBuilder` is more convenient and does not introduce too much maintenance burden.

cc @FANNG1

**Comment by FANNG1** [comment]
How about Override `loadTable(Identifier ident, String version)` and `loadTable(Identifier ident, String version)` for `GravitinoIcebergCatalog`?

**Comment by caican00** [comment]
> How about Override `loadTable(Identifier ident, String version)` and `loadTable(Identifier ident, String version)` for `GravitinoIcebergCatalog`?

in this way, we still have the problem of initializing the `snapshotId` or `branch` when invoking `super` method of `SparkTable`.

**Comment by FANNG1** [comment]
> > How about Override `loadTable(Identifier ident, String version)` and `loadTable(Identifier ident, String version)` for `GravitinoIcebergCatalog`?
> 
> in this way, we still have the problem of initializing the `snapshotId` or `branch` when invoking `super` method of `SparkTable`.

Let me think a while

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/SparkHiveTable.java [hunk 1]
```diff
@@ -55,4 +59,10 @@ public Map<String, String> properties() {
   public Transform[] partitioning() {
     return gravitinoTableInfoHelper.partitioning();
   }
+
+  /** to keep consistent behavior with SparkIcebergTable. */
```
Comment (FANNG1) [inline spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/SparkHiveTable.java:63]: I think there's no need to keep consistent with Iceberg

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java [hunk 2]
```diff
@@ -40,23 +39,18 @@ protected TableCatalog createAndInitSparkCatalog(
     return icebergCatalog;
   }
 
+  /**
+   * Migrated `loadTable(identifier)` to the BaseCatalog class and execute `loadTable(identifier)`
+   * before createSparkTable to load sparkTable with different parameters easily.
+   */
```
Comment (FANNG1) [inline spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java:45]: this is more like a comment in PR not in code

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/catalog/BaseCatalog.java [hunk 3]
```diff
@@ -206,14 +214,52 @@ public Table createTable(
...
+          sparkCatalog,
+          propertiesConverter,
+          sparkTransformConverter);
+    } catch (com.datastrato.gravitino.exceptions.NoSuchTableException e) {
+      throw new NoSuchTableException(ident);
+    }
+  }
+
+  @Override
+  public Table loadTable(Identifier ident, String version) throws NoSuchTableException {
```
Comment (FANNG1) [inline spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/catalog/BaseCatalog.java:233]: better to place these methods to SparkIcebergCatalog

Comment (caican00) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/catalog/BaseCatalog.java:233]: in this way, paimon need to implement these methods again.

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java (hunk 2 above)
Comment (caican00) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java:45]: should i remove it?

Comment (FANNG1) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java:45]: ok

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/catalog/BaseCatalog.java (hunk 3 above)
Comment (FANNG1) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/catalog/BaseCatalog.java:233]: we could provide something like `TimetravelCatalog` when supporting paimon, leaving BaseCatalog do basic operations.

Comment (caican00) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/catalog/BaseCatalog.java:233]: ok.

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/SparkHiveTable.java (hunk 1 above)
Comment (caican00) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/SparkHiveTable.java]: updated

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java (hunk 2 above)
Comment (caican00) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java]: updated.

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/catalog/BaseCatalog.java (hunk 3 above)
Comment (caican00) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/catalog/BaseCatalog.java]: done

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/SparkIcebergTable.java [hunk 4]
```diff
@@ -59,6 +63,16 @@ public Transform[] partitioning() {
     return gravitinoTableInfoHelper.partitioning();
   }
 
+  /**
+   * Although SparkIcebergTable extended SparkTable, it also needs to initialize its member variable
+   * , such as snapshotId or branch, before it reused newScanBuilder from the parent class. In
+   * contrast, overriding newScanBuilder to support time travel is simpler and more concise.
+   */
+  @Override
+  public ScanBuilder newScanBuilder(CaseInsensitiveStringMap options) {
```
Comment (FANNG1) [inline spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/SparkIcebergTable.java:72]: could you rewrite comment like this?
```
If using SparkIcebergTable not SparkTable, we must extract snapshotId or branchName using the Iceberg specific logic. It's hard to maintenance.
```

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java [hunk 5]
```diff
@@ -239,6 +243,75 @@ void testIcebergTableRowLevelOperations(IcebergTableWriteProperties icebergTable
...
+    Assertions.assertEquals(1, tableData.size());
+    Assertions.assertEquals("1,1,1", tableData.get(0));
+
+    SparkIcebergTable sparkIcebergTable = getSparkIcebergTableInstance(tableName);
+    long snapshotId = getCurrentSnapshotId(tableName);
+    sparkIcebergTable.table().manageSnapshots().createBranch("test_branch", snapshotId).commit();
+    sparkIcebergTable.table().manageSnapshots().createTag("test_tag", snapshotId).commit();
+    long snapshotTimestamp = getCurrentSnapshotTimestamp(tableName);
+    long timestamp = waitUntilAfter(snapshotTimestamp + 1000);
+    waitUntilAfter(timestamp + 1000);
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java:264]: `waitUntilAfter(timestamp + 1000);`  should be removed?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java [hunk 6]
```diff
@@ -239,6 +243,75 @@ void testIcebergTableRowLevelOperations(IcebergTableWriteProperties icebergTable
     testIcebergMergeIntoUpdateOperation(icebergTableWriteProperties);
   }
 
+  @Test
+  void testIcebergAsOfQuery() throws NoSuchTableException {
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java:247]: testIcebergTimeTravelQuery?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java [hunk 7]
```diff
@@ -559,4 +632,32 @@ static IcebergTableWriteProperties of(
       return new IcebergTableWriteProperties(isPartitionedTable, formatVersion, writeMode);
     }
   }
+
+  private SparkIcebergTable getSparkIcebergTableInstance(String tableName)
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java:636]: could we get `snapshotId` or `SnapshotTimestamp` from SparkSQL not internal API?

Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java:636]: we have not support metadata tables yet, so we can not get `snapshotId` or `SnapshotTimestamp` from SparkSQL.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java (hunk 6 above)
Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java:247]: ok.

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/SparkIcebergTable.java (hunk 4 above)
Comment (caican00) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/SparkIcebergTable.java:72]: ok

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java (hunk 7 above)
Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java:636]: i think we can refactor it when we support metadata tables.

Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java:636]: got it

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/SparkIcebergTable.java (hunk 4 above)
Comment (caican00) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/SparkIcebergTable.java:71]: updated.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java (hunk 5 above)
Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java]: ok. done.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java (hunk 6 above)
Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java]: updated.

**Comment by caican00** [comment]
fixed conflict, and could you please help review again if you are free? Thank you.

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java [hunk 8]
```diff
@@ -136,4 +166,42 @@ private boolean isSystemNamespace(String[] namespace)
     isSystemNamespace.setAccessible(true);
     return (Boolean) isSystemNamespace.invoke(baseCatalog, (Object) namespace);
   }
+
+  private com.datastrato.gravitino.rel.Table loadGravitinoTable(Identifier ident)
```
Comment (FANNG1) [inline spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java:170]: could you place this method in `baseCatalog` and resued by `loadTable`?

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/catalog/BaseCatalog.java [hunk 9]
```diff
@@ -65,9 +65,9 @@ public abstract class BaseCatalog implements TableCatalog, SupportsNamespaces {
   protected Catalog gravitinoCatalogClient;
   protected PropertiesConverter propertiesConverter;
   protected SparkTransformConverter sparkTransformConverter;
+  protected final String metalakeName;
```
Comment (FANNG1) [inline spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/catalog/BaseCatalog.java:68]: no need to protected if moving `loadGravitinoTable`?

**Comment by FANNG1** [comment]
LGTM, except minor comments

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java (hunk 8 above)
Comment (caican00) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java]: done.

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/catalog/BaseCatalog.java (hunk 9 above)
Comment (caican00) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/catalog/BaseCatalog.java]: yes. have rolled it back.

**Comment by caican00** [comment]
> LGTM, except minor comments

comments have been addressed, and could you please help review again? @FANNG1

**Review by FANNG1 — APPROVED** [review]

**Comment by FANNG1** [comment]
@caican00 , thanks for your contribution!

**Pull request merged**